libs/core/src/main/java/org/opensearch/common/SetOnce.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.common;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A convenient class which offers a semi-immutable object wrapper implementation which allows one
+ * to set the value of an object exactly once, and retrieve it many times. If {@link #set(Object)}
+ * is called more than once, {@link AlreadySetException} is thrown and the operation will fail.
+ *
+ * This is borrowed from lucene's experimental API. It is not reused to eliminate the dependency
+ * on lucene core for such a simple (standalone) utility class that may change beyond OpenSearch needs.
+ *
+ * @opensearch.internal
```

Comment:
Aren't these public utilities in the core library fair game to be used by plugins? Several of the plugins packaged in this repo are using this.

Reply:
yes! Good call out. I'll change the label to `@opensearch.api`

Reply:
Nothing in `libs/core` currently has any of these javadoc annotations. Is it fair to say that everything Java `public` in a library module is part of the API, or do we want to start tagging things with annotations?

Reply:
> Nothing in libs/core currently has any of these javadoc annotations.

I added them in #5902 (along with enabling `missingJavadocs` task for lib-core) but started teasing that change into a separate PR. 


> Is it fair to say that everything Java public in a library module is part of the API

That would be a reasonable assumption but alas, the inherited code did not thoughtfully apply java modifiers. The default behavior of the predecessor was `public`. I'm not sure what the best course would be for this beyond "proceed at your own risk" at which point maybe start conservatively `@opensearch.internal` (like my change did) and address on a per class basis? I'm open to majority opinion on this.

Reply:
Maybe I'm being naive, but I would think that a library like `:core` would contain few to none of the cross package dependencies of logically internal pieces like is common in `:server`, which means we could rely on the Java access qualifiers. If there is anything currently `public` that should not be, then we could `@Deprecate` it now and remove/change qualifier in 3.0. Is that tenable?

Reply:
> I would think that a library like :core would contain few to none of the cross package dependencies of logically internal pieces like is common in :server

Correct. There shouldn't be any cross package dependencies. `:server` should depend on `opensearch-core` and no dependency the other way around.

Furthermore, IIRC, the original intent was to have zero dependencies in `opensearch-core`. I'm not sure what the benefit of that really is (if anything it restricts what can be refactored to core) so #5902 breaks that intent by introducing a dependency on the jackson library (which used to be a dependency in `x-content`). We could continue to discuss whether we want/need a dependency free `commons` library in the #5902 PR and if so consider a new dependency free library called `opensearch-commons` and let `opensearch-core` bring in dependencies.